src/ui/public/routes/route_setup_manager.js

```diff
@@ -71,9 +73,18 @@ module.exports = class RouteSetupManager {
 
       return defer.promise.then(() => Promise.all(userWork.doWork()));
     })
+    .catch(error => {
+      if (error !== WAIT_FOR_URL_CHANGE_TOKEN) {
+        throw error;
+      }
+
+      // prevent moving forward, return a promise that never resolves
+      // so that the $router can observe the $location update
+      return Promise.halt();
```

Comment:
Super minor nit, but this seems more like the "early exit" exception, and throwing the error seems more like the default flow, so I'd suggest swapping the two:

```javascript
if (error === WAIT_FOR_URL_CHANGE_TOKEN) {
  // prevent moving forward, return a promise that never resolves
  // so that the $router can observe the $location update
  return Promise.halt();
}

throw error;
```